core/src/main/java/org/infinispan/reactive/publisher/impl/LocalPublisherManagerImpl.java

```diff
@@ -414,14 +419,18 @@ private int getNextSegment(PrimitiveIterator.OfInt segmentIter) {
          Function<? super Publisher<R>, ? extends CompletionStage<R>> finalizer) {
       AdvancedCache<K, V> cache = deliveryGuarantee == DeliveryGuarantee.AT_MOST_ONCE ? this.cache : remoteCache;
       return handleSpecificObjects(parallelPublisher, keysToInclude, keysToExclude, keyFlowable ->
-         keyFlowable.map(k -> {
-            CacheEntry<K, V> entry = cache.getCacheEntry(k);
-            if (entry == null) {
-               return NullCacheEntry.<K, V>getInstance();
-            }
-            return entry;
-         }).filter(e -> e != NullCacheEntry.getInstance())
-      , collator, finalizer);
+                  keyFlowable.concatMapMaybe(k -> {
+                     CompletableFuture<CacheEntry<K, V>> future = cache.getCacheEntryAsync(k);
+                     future.thenApply(entry -> {
+                        if (entry instanceof MVCCEntry) {
+                           // Scattered cache can return MVCCEntry instances
+                           entry = new ImmortalCacheEntry(entry.getKey(), entry.getValue());
+                        }
+                        return entry;
+                     });
+                     return RxJavaInterop.completionStageToMaybe(future);
```

Comment:
Don't we need to pass the `CompletableFuture` returned by `thenApply` to `RxJavaInterop.completionStageToMaybe`?

Reply:
Yes.